lib/engine/game/g_1812/game.rb

```diff
@@ -139,58 +396,118 @@ def can_par?(corporation, parrer)
         end
 
-        NORTH_HEXES = %w[A4 A8 F1].freeze
-        SOUTH_HEXES = %w[C20 E20 F19].freeze
-        PORT_HEXES = %w[F3 G4 G6 G8 H9 H17 H19].freeze
-        MINE_HEXES = %w[B15 D7 D17 E2 E6].freeze
-        GTRAINS = %w[1G 2G 2+1G 3+2G 4+2G 2+2GD].freeze
-        F3_PORT = ['F3'].freeze
-        G6_PORT = ['G6'].freeze
-        H9_PORT = ['H9'].freeze
+        def unpaid_loan(entity, owed)
+          current_cash = entity.cash
+          old_price = entity.share_price
+
+          @log << "#{entity.name} owes #{format_currency(owed)} in loan interest but has #{format_currency(entity.cash)}"
+          @log << "#{entity.name} will pay all remaining treasury cash (#{format_currency(current_cash)}) to the "\
+                  'bank and its stock price will drop one space'
+          entity.spend(current_cash, bank) if current_cash.positive?
+          @stock_market.move_left(entity)
+          log_share_price(entity, old_price)
+        end
 
-        def mine_port_bonus
-          @hexes.find { |hex| hex.coordinates == 'I3' }.tile.offboards.first
+        def trainless_penalty(entity)
+          old_price = entity.share_price
+
+          @log << "#{entity.name} was unable to buy a train and is trainless. Its stock price will drop one space"
+          @stock_market.move_left(entity)
+          log_share_price(entity, old_price)
+        end
+
+        G_TRAINS = %w[1G 2G 2+1G 3+2G 4+2G 2+2GD].freeze
+        PORT_HEXES = %w[F3 G4 G6 G8 H9 H17 H19].freeze
+        NORTH_SOUTH_BONUS_HEX = 'I1'
+        PORT_MINE_BONUS_HEX = 'I3'
+        C18_HEX = 'C18'
+        F3_PORT = 'F3'
+        G6_PORT = 'G6'
+        H9_PORT = 'H9'
+
+        def g_train?(train)
+          self.class::G_TRAINS.include?(train.name)
         end
 
-        def ns_bonus
-          @hexes.find { |hex| hex.coordinates == 'I1' }.tile.offboards.first
+        def check_other(route)
+          return if g_train?(route.train)
+          return unless (route.stops.map(&:hex).map(&:id) & PORT_HEXES).any?
+
+          raise GameError, 'Only G trains can run to ports'
         end
 
         def revenue_for(route, stops)
           revenue = super
-          hex = route.hexes
-          gtrain = route.train.variant.name?(GTRAINS)
+          train = route.train
+          corp = route.corporation
+
+          revenue += 10 if corp.assigned?(p4_company) && stops.find { |s| C18_HEX.include?(s.hex.id) }
+
+          revenue += 10 if g_train?(train) && corp.assigned?(p3_company) && stops.any? { |s| F3_PORT.include?(s.hex.id) }
+          revenue += 10 if g_train?(train) && corp.assigned?(p8_company) && stops.any? { |s| F3_PORT.include?(s.hex.id) }
+          revenue += 10 if g_train?(train) && corp.assigned?(p9_company) && stops.any? { |s| H9_PORT.include?(s.hex.id) }
+          revenue += 20 if g_train?(train) && corp.assigned?(p12_company) && stops.any? { |s| G6_PORT.include?(s.hex.id) }
```

Comment:
This very repetitive. Could you make a method that encapsulates these tests?

Reply:
Actually, I think I got something. Let me know what you think. Tbh, I feel like it looks more cluttered. The reason I had it as separate lines was they were all clean and easily read/understood.